### PR TITLE
Torch export support

### PR DESCRIPTION
**Pull request by pctablet505**
## Summary

This PR adds two new model export capabilities to Keras:

1. **`format="torch"` — PyTorch ExportedProgram (`.pt2`) export**  
   Export any Keras model (torch backend) to a portable `.pt2` artifact via `torch.export.export`. The file can be loaded in any PyTorch environment without Keras.

2. **LiteRT export via the Torch backend**  
   When `KERAS_BACKEND=torch`, `model.export(format="litert")` now invokes `litert-torch` instead of going through a TensorFlow SavedModel. This eliminates Flex-delegate issues caused by ops that TF traces but TFLite doesn't support natively.

Additionally, several correctness and compatibility fixes are included for the PyTorch backend's slice operations and the SavedModel export archive.

---

### Problem 1 — No stand-alone `.pt2` export

Keras had no way to produce a `torch.export.ExportedProgram` artifact. Users who wanted to deploy Keras (torch-backend) models with `litert-torch` or directly in a PyTorch serving stack had to manually call `torch.export.export` themselves.

### Problem 2 — Flex-delegate fallbacks in LiteRT via TF path

`model.export(format="litert")` with the TF backend pipes through a SavedModel intermediate. Operations like `tf.StridedSlice(new_axis_mask=...)` (produced by Python `None`-indexing) and `HashTable` lookup ops fall back to the Flex delegate, which is not available in standalone `ai_edge_litert ≥ 2.20`. Using `litert-torch` directly avoids the TF graph entirely.

### Problem 3 — `slice_update` / `dynamic_slice` break `torch.export`

When `start_indices` was a Python list the old code still called `convert_to_tensor`, creating data-dependent tensor expressions that `torch.export.export` cannot trace symbolically.

---

### Export flow overview

```mermaid
flowchart TD
    A[model.export] --> B{format?}
    B -- litert --> C{backend?}
    B -- torch --> D[TorchExporter]
    C -- TF --> E[LiteRTExporter]
    C -- torch --> F[export_litert_via_torch]
    D --> G[torch.export.export]
    G --> H[.pt2 artifact]
    E --> I[.tflite artifact]
    F --> I
```

### New `TorchExporter` class

```mermaid
classDiagram
    class TorchExporter {
        +model : keras.Model
        +input_signature
        +kwargs : dict
        +export(filepath) str
        -_get_model_device() device
        -_get_export_kwargs() dict
    }

    class export_torch {
        <<function>>
        +model
        +filepath
        +input_signature
        +verbose
    }

    export_torch --> TorchExporter : instantiates
    TorchExporter --> torch_export_export : calls
```

### LiteRT dual-backend routing

```mermaid
flowchart LR
    subgraph litert.py
        A[export_litert] --> B{backend}
        B -- torch --> C[export_litert_via_torch]
        B -- tensorflow --> D[LiteRTExporter]
        C --> E[litert_torch.convert]
        E --> F[litert_torch.save]
        D --> G[SavedModel]
        G --> H[tf.lite.TFLiteConverter]
        H --> I[.tflite]
        F --> I
    end
```

### TF Backend: LiteRT Export Pipeline Detail

```mermaid
flowchart TD
    A[LiteRTExporter.export] --> B[Validate .tflite extension]
    B --> C[_convert_to_tflite]
    C --> D[Create ExportArchive]
    D --> E["archive.add_endpoint('serving_default', model.call)"]
    E --> F[archive.write_out to temp SavedModel dir]
    F --> G[TFLiteConverter.from_saved_model]
    G --> H[_apply_converter_kwargs]
    H --> I{Attempt 1: TFLITE_BUILTINS only}
    I -- success --> J[Return .tflite bytes]
    I -- fail --> K{Attempt 2: BUILTINS + SELECT_TF_OPS}
    K -- success --> L[Emit Flex delegate warning]
    L --> J
    K -- fail --> M[RuntimeError]
```

The TF backend pipeline avoids two Keras 3 incompatibilities:

1. **Cannot use `TFLiteConverter.from_keras_model`** — calls `keras_deps.get_call_context_function()` which returns `None` in Keras 3 and crashes
2. **Cannot use `tf.saved_model.save` directly** — calls `inspect.getattr_static(obj, '__dict__')` on Keras 3's `_DictWrapper`, raising `TypeError`

The solution uses `ExportArchive.add_endpoint()` which traces `model.call` directly and captures only the concrete function's variables, bypassing both issues.

**Two-pass conversion strategy:**
- **Pass 1**: `TFLITE_BUILTINS` only — produces a model that runs without the Flex delegate (ideal for `ai_edge_litert` standalone runtime)
- **Pass 2** (fallback): `TFLITE_BUILTINS + SELECT_TF_OPS` — emits a `UserWarning` so the caller knows the Flex delegate will be needed at inference time
- If both passes fail, raises `RuntimeError` with the original exception

### Torch Backend: `export_litert_via_torch` Pipeline Detail

```mermaid
flowchart TD
    A[export_litert_via_torch] --> B["Step 1: _move_model_to_cpu\n(3 tensor categories)"]
    B --> C["Step 2: _register_litert_decompositions\n(MPS SDPA, mean.dim, repeat_interleave)"]
    C --> D["Step 3: _patch_vhlo_target_version\n(StableHLO compat)"]
    D --> E["Step 4: Prepare sample_inputs\nfrom input_signature"]
    E --> F["Step 5: model.eval()"]
    F --> G["Step 6: litert_torch.convert\n+ edge_model.export"]
    G --> H["Step 7: _restore_model_devices"]
```

**Step 1 — CPU migration** handles three tensor categories:
- Keras `Variable` instances (`model.variables`)
- `nn.Module` parameters and buffers
- Raw tensor attributes on layers (e.g. `Normalization.mean` / `.variance`)

**Step 2 — Decompositions** registers portable replacements for ops that `litert-torch` cannot lower:
- `_scaled_dot_product_attention_math_for_mps` (MPS-only SDPA op)
- `aten.mean.dim` with `dtype` kwarg (strips dtype, decomposes via sum/count)
- `aten.repeat_interleave.Tensor` (decomposed via `cumsum` + `searchsorted`)
- `aten.repeat_interleave.self_int` (decomposed via `unsqueeze` + `expand` + `reshape`)

**Step 3 — VHLO patching** monkey-patches `MlirLowered.module_bytecode_vhlo` to:
- Remove `stablehlo.optimization_barrier` ops (compiler hints with no runtime effect)
- Serialize at minimum StableHLO version (avoiding v1.13+ ops like `tanh_v2` that the TFLite converter cannot parse)

**Step 6 — Quantization translation** via `_prepare_litert_kwargs()`:
- Translates TFLite-style `tf.lite.Optimize.*` constants to `litert_torch` `QuantConfig`
- `DEFAULT` maps to dynamic-range quantization
- `OPTIMIZE_FOR_SIZE` / `OPTIMIZE_FOR_LATENCY` maps to static per-channel quantization via `PT2EQuantizer`

### Quantization support comparison

```mermaid
flowchart LR
    subgraph TF Backend
        A[User kwargs] --> B[_apply_converter_kwargs]
        B --> C[converter.optimizations]
        B --> D[converter.representative_dataset]
        B --> E[converter.target_spec.*]
    end
    subgraph Torch Backend
        F[User kwargs] --> G[_prepare_litert_kwargs]
        G --> H[_create_quant_config_from_optimizations]
        H --> I[PT2EQuantizer + QuantConfig]
    end
```

| Feature | TF Backend | Torch Backend |
|---|---|---|
| Dynamic range quantization | `converter.optimizations = [DEFAULT]` | `QuantConfig(pt2e_quantizer)` with `is_dynamic=True` |
| Static per-channel | `converter.optimizations = [OPTIMIZE_FOR_SIZE]` | `QuantConfig` with `is_per_channel=True` |
| Full-integer (INT8) | `representative_dataset` kwarg | `quant_config` kwarg (direct) |
| Flex delegate opt-in | Automatic fallback (two-pass) | N/A (no TF ops in torch pipeline) |
| Custom converter kwargs | Any `converter.*` attribute via `**kwargs` | `strict_export`, `dynamic_shapes`, internal flags |

---

### `keras/src/export/torch.py` *(NEW — 200 lines)*

New module introducing:

| Symbol | Purpose |
|---|---|
| `export_torch(model, filepath, ...)` | Public API function, called by `model.export(format="torch")` |
| `TorchExporter` | Encapsulates `torch.export.export` + `torch.export.save` |
| `TorchExporter._get_model_device()` | Detects device from model variables |
| `TorchExporter._get_export_kwargs()` | Merges user kwargs, defaults `strict=False` |

Key design decisions:
- `strict=False` by default — handles most Keras models without manual graph annotation
- Input signature inferred via `get_input_signature()` when not provided
- `None` batch dimensions replaced with `1` for concrete sample tensors
- Comprehensive error message on export failure (data-dependent control flow, etc.)

### `keras/src/export/litert.py` *(+774 lines, -105 lines)*

This is the largest changed file. It contains both the refactored TF backend exporter and the entirely new torch backend pipeline.

#### `export_litert()` — Top-level router

```python
def export_litert(model, filepath, input_signature=None, verbose=None, **kwargs):
    if backend.backend() == "torch":
        return export_litert_via_torch(model, filepath, ...)
    if backend.backend() != "tensorflow":
        raise ImportError(...)
    exporter = LiteRTExporter(model=model, input_signature=input_signature, **kwargs)
    exporter.export(filepath)
```

#### `LiteRTExporter` class — TF backend pipeline

| Method | Purpose |
|---|---|
| `__init__(model, input_signature, **kwargs)` | Stores model, signature, and user kwargs |
| `export(filepath)` | Validates `.tflite` extension, calls `_convert_to_tflite()`, writes bytes |
| `_convert_to_tflite(input_signature)` | Core pipeline: ExportArchive trace, SavedModel, TFLiteConverter (two-pass) |
| `_apply_converter_kwargs(converter)` | Passes user kwargs to TFLiteConverter (handles nested `target_spec` dict) |

The `_convert_to_tflite` method is key — it uses `ExportArchive.add_endpoint("serving_default", model.call)` to trace the model into a temp SavedModel, then converts via `TFLiteConverter.from_saved_model()`. The two-pass approach tries `TFLITE_BUILTINS` first, then falls back to `TFLITE_BUILTINS + SELECT_TF_OPS` with a warning.

#### `export_litert_via_torch()` — Torch backend pipeline (new)

| Symbol | Purpose |
|---|---|
| `export_litert_via_torch()` | 7-step pipeline: CPU move, decompositions, VHLO patch, convert, restore |
| `_move_model_to_cpu()` | Handles Keras Variables, nn.Module params/buffers, raw tensor attrs |
| `_restore_model_devices()` | Restores all tensors to original devices after export |
| `_register_litert_decompositions()` | Portable replacements for MPS SDPA, mean.dim, repeat_interleave |
| `_patch_vhlo_target_version()` | Monkey-patches VHLO serialization for StableHLO version compat |
| `_remove_optimization_barriers()` | Walks MLIR, strips `stablehlo.optimization_barrier` ops |
| `_prepare_litert_kwargs()` | Translates TFLite-style `optimizations` to `litert_torch` `QuantConfig` |
| `_create_quant_config_from_optimizations()` | Maps `tf.lite.Optimize.*` to `PT2EQuantizer` configs |

Quantization support added via both paths:
- `DEFAULT` / `FLOAT16` / `INT8` optimization levels
- `representative_dataset` for full-integer quantization
- `enable_select_tf_ops` for Flex delegate opt-in

### `keras/src/models/model.py` *(+51, -5)*

```python
# Before
EXPORT_FORMATS = ("tf_saved_model", "litert")

# After  
EXPORT_FORMATS = ("tf_saved_model", "litert", "torch")
```

The `export()` method now:
1. Validates `"torch"` is only used with `KERAS_BACKEND=torch`
2. Routes `format="torch"` to `export_torch()`
3. Accepts `format="litert"` on both TF and torch backends

### `keras/src/export/saved_model_export_archive.py` *(+11, -5)*

**Bug fix:** `SavedModelTrackableView.children()` raised `TypeError` when TF's `convert_to_trackable` calls `inspect.getattr_static(obj, '__dict__')` on Keras 3's `_DictWrapper` type. The subclass now overrides `children()` with `save_type="savedmodel"` (TF hardcodes `"checkpoint"`) and wraps in `try/except TypeError`:

```python
class SavedModelTrackableView(tf.train.TrackableView):
    @classmethod
    def children(cls, obj, save_type="savedmodel", **kwargs):
        try:
            return super().children(obj, save_type, **kwargs)
        except TypeError:
            # Keras 3 _DictWrapper raises TypeError via
            # inspect.getattr_static(obj, '__dict__').
            # Variables are already tracked via _list_variables_used_by_fns;
            # only TrackableResource assets on this subtree would be missed.
            return {}
```

This is safe because variables used by endpoints are already captured separately via `_list_variables_used_by_fns()`. Only `TrackableResource` assets (lookup tables) on the skipped subtree could be missed, which is acceptable for LiteRT export.

### `keras/src/backend/torch/core.py` *(+46, -14)*

`dynamic_slice` and `slice_update` now have a **fast path** for static Python lists:

```python
# Fast path — avoids data-dependent expressions that break torch.export
if isinstance(start_indices, (list, tuple)) and all(isinstance(s, int) for s in start_indices):
    slices = [slice(start, start + length) for start, length in zip(start_indices, shape)]
    return inputs[slices]

# Slow path — uses torch.narrow for truly dynamic tensor indices
result = inputs
for dim in range(start_indices.shape[0]):
    result = torch.narrow(result, dim, start_indices[dim], shape[dim])
return result
```

### `keras/src/backend/torch/nn.py` *(+9)* and `keras/src/backend/torch/numpy.py` *(+18)*

Minor compatibility fixes for ops used within models that are exported:
- `nn.py`: Guard for `torch.nn.functional` ops that require contiguous tensors
- `numpy.py`: `cumsum`/`cumprod` dtype handling for `torch.export` traceability

---

### `keras/src/export/torch_test.py` *(NEW — 415 lines)*

| Test Class | Coverage |
|---|---|
| `TorchExportBasicTest` | Single-input dense models, roundtrip load+inference |
| `TorchExportInputSignatureTest` | Explicit vs inferred signatures |
| `TorchExportConvTest` | 2D conv models with image inputs |
| `TorchExportMultiInputTest` | Dict-input models |
| `TorchExportErrorTest` | Error messages, wrong backend, bad filepath |
| `TorchExportQuantTest` | Export after `torch.compile` |

### `keras/src/export/litert_torch_test.py` *(NEW — 479 lines)*

| Test Class | Coverage |
|---|---|
| `LiteRTViaTorchBasicTest` | Dense model export + inference via `litert-torch` |
| `LiteRTViaTorchMultiInputTest` | Dict-input model (attention-like) |
| `LiteRTViaTorchConvTest` | Conv2D model with image inputs |
| `LiteRTViaTorchNumericsTest` | Output value comparison Keras ↔ LiteRT |
| `LiteRTViaTorchDynamicShapeTest` | Known-failing data-dependent shapes (xfail) |

### `keras/src/export/litert_test.py` *(+45 lines)*

Added tests for quantization levels (`DEFAULT`, `FLOAT16`, `INT8`) and the new backend-routing logic.

---

## Code Review Questions

1. **`strict=False` default**: `torch.export.export(strict=False)` is used so Keras models (which have dynamic control flow from `training=True/False`) can be traced. Should this be configurable at the `model.export()` API level (e.g., `model.export("m.pt2", format="torch", strict=True)`)?

2. **`SavedModelTrackableView.children()` fix**: The `try/except TypeError` in `saved_model_export_archive.py` is a defensive patch against Keras 3's `_DictWrapper`. Is this the right place for this guard, or should `_DictWrapper` itself expose `_trackable_children(cache=...)`?

3. **`export_litert_via_torch` signature**: Should `export_litert_via_torch` accept the same quantization kwargs as `LiteRTExporter`, forwarding them to `litert-torch`'s quantization API? Currently it only does fp32 export.

4. **`None` batch dimension replacement**: `TorchExporter` replaces `None` dims with `1` for sample inputs. Should it be `2` (to catch some batch-norm edge cases) or allow the user to specify a `default_batch_size` kwarg?

5. **`.pt2` format extension enforcement**: We currently raise `ValueError` if the filepath doesn't end in `.pt2`. Is this the right UX, or should we silently append `.pt2`?

---

## Testing

```bash
# Torch export tests
KERAS_BACKEND=torch pytest keras/src/export/torch_test.py -v

# LiteRT via torch tests
KERAS_BACKEND=torch pytest keras/src/export/litert_torch_test.py -v

# Full LiteRT tests (TF backend)
KERAS_BACKEND=tensorflow pytest keras/src/export/litert_test.py -v

# SavedModel export tests
KERAS_BACKEND=tensorflow pytest keras/src/export/saved_model_test.py -v
```

Expected results (torch backend):
- `torch_test.py`: all pass
- `litert_torch_test.py`: all pass except dynamic-shape tests (xfail)

---

## Dependency Notes

| Package | Required for |
|---|---|
| `torch` | `format="torch"` and `format="litert"` on torch backend |
| `litert-torch` | LiteRT export via torch backend (`litert_torch.convert()`) |
| `litert-torch` | LiteRT inference from `.tflite` on torch backend |
| `ai-edge-litert` | LiteRT inference on TF backend |

These are optional extras — missing packages produce a clear `ImportError` with installation instructions.

### CODE REVIEW

**Comment by pctablet505** [comment]
@gemini-code-assist review

**Comment by pctablet505** [comment]
@gemini-code-assist review

**Comment by pctablet505** [comment]
@gemini-code-assist review

**Comment by pctablet505** [comment]
@gemini-code-assist review

**Comment by hertschuh** [comment]
This PR is in Draft, let me know when it's ready for review.

**Comment by pctablet505** [comment]
https://github.com/keras-team/keras/pull/22758 this is new PR
